core/src/main/java/com/linecorp/armeria/client/retry/RetryRule.java

```diff
@@ -0,0 +1,210 @@
+/*
+ * Copyright 2020 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.client.retry;
+
+import java.util.concurrent.CompletionStage;
+import java.util.function.Predicate;
+
+import javax.annotation.Nullable;
+
+import com.linecorp.armeria.client.ClientRequestContext;
+import com.linecorp.armeria.client.UnprocessedRequestException;
+import com.linecorp.armeria.common.HttpStatus;
+import com.linecorp.armeria.common.HttpStatusClass;
+import com.linecorp.armeria.common.Response;
+import com.linecorp.armeria.common.ResponseHeaders;
+
+/**
+ * Determines whether a failed request should be retried.
+ * If you need to determine by looking into the {@link Response}, use {@link RetryRuleWithContent}.
+ */
+@FunctionalInterface
+public interface RetryRule {
+
+    /**
+     * Returns a newly created {@link RetryRule} that will retry with the
+     * {@link Backoff#ofDefault() default backoff} if the request HTTP method is
+     * <a href="https://developer.mozilla.org/en-US/docs/Glossary/Idempotent">idempotent</a>
+     * and an {@link UnprocessedRequestException} is raised or the class of the response status is
+     * {@link HttpStatusClass#SERVER_ERROR}.
+     *
+     * <p>Note that a client can safely retry a failed request with this rule if an endpoint service produces
+     * the same result (no side effects) on idempotent HTTP methods.
+     *
+     * <p>This method is shortcut for:
+     * <pre>{@code
+     * RetryRule.builder()
+     *          .onIdempotentMethods()
+     *          .onServerErrorStatus()
+     *          .onUnprocessed()
+     *          .thenBackoff();
+     * }</pre>
+     */
+    static RetryRule failsafe() {
+        return builder().onIdempotentMethods().onServerErrorStatus().onUnprocessed().thenBackoff();
```

Comment:
I think @minwoox's suggestion of using `or` between unprocessed and idempotent.serverErrorStatus makes sense. For this rule, is it even possible to retry? I guess an unprocessed request should never have a server error status since it didn't make it to the server?

Reply:
Wondering maybe we can just remove `onUnprocessed` and have `RetryingClient` always retry on unprocessed regardless of rule. Is there any conceivable case not to retry on unprocessed (from what I understand unprocessed means not sent to the server, but the only reason clients exist in the first place is solely to send something to the server and should always want to IMO). Putting it this way, I wonder if we even should require `RetryingClient` to retry unprocessed or just do it natively.

Reply:
> I guess an unprocessed request should never have a server error status since it didn't make it to the server?

You are right. `on*` methods in `RetryRuleBuilder` is OR expression except `on*Methods`.

`builder().onIdempotentMethods().onServerErrorStatus().onUnprocessed().thenBackoff()` means
```
rule = request method is idempotent && (response status is 5xx || cause is UnprocessedRequestException)
```
Does this make you confused? Hmm...

The main reason why I add `or()` method is to allow a user configuring different backoff for different errors.

> Wondering maybe we can just remove onUnprocessed and have RetryingClient always retry on unprocessed regardless of rule.

That makes sense, but we maybe need a discussion for this topic. Let me file a issue soon.

Reply:
Hmm - yeah same type of methods having different or / and behavior seems confusing, you really need to know this API to be able to reason about such code. I'm reviewing the PR and didn't realize ;) Can't come up with a good idea, which is the main reason for the suggestion of just always retrying unprocessed.

Reply:
I think `AND` operation in builder does not make sense for a response status or an exception.
How about changing `onIdempotentMethods()` to `ifIdempotentMethods()`?

```java
RetryRule.builder()
         .ifIdempotentMethods()
         .onServerErrorStatus()
         .onUnprocessed()
         .thenBackoff()
         .or(RetryRule.onStatus(HttpStatus.TOO_MANY_REQUESTS));
```

or just give up changing `on*` methods and 
We can change the return type builder methods to select only one `on*` methods.

```java
class RetryRuleBuilder {
    public RetryBackoffBuilder onServerErrorStatus() { 
        return new RetryBackoffBuilder(this); 
    }
}
class RetryBackoffBuilder {

   public RetryRule thenBackoff(Backoff backoff) { 
       ...
   } 
}
```

Reply:
If is a little better, but if some user code is on, if, on, it looks a bit weird.

Maybe we can make methods a constructor parameter / factory parameter for the builder? Then it's clearer they're special

Reply:
It sounds good mostly but can't offer special method set such as `idempotentMethods()`.

Then probably need a special builder method?
```java
RetryRule.builder()
         .onServerErrorStatus()...

RetryRule.builder(HttpMethod.GET)
         .onServerErrorStatus()...

RetryRule.builderWithIdempotentMethods()
         .onServerErrorStatus()...
```

Reply:
Yeah that's what I was thinking

Reply:
How about `RetryRule.builder(HttpMethod.idempotentMethods())`?